allennlp/data/iterators/data_iterator.py

```diff
@@ -0,0 +1,49 @@
+from typing import Dict, List, Generator
+
+import numpy
+
+from allennlp.data import Dataset
+from allennlp.common import Params
+
+
+class DataIterator:
+    """
+    An abstract ``DataIterator`` class. ``DataIterators`` must implement __call__, which yields
+    batched examples.
+    """
+    def __call__(self, dataset: Dataset) -> Generator[Dict[str, List[numpy.array]], None, None]:
+        """
+        Returns a generator that yields batches over the given dataset, forever.
+
+        Parameters
+        ----------
+        dataset : ``Dataset``
+        """
+        raise NotImplementedError
+
+    def yield_one_pass(self, dataset: Dataset) -> Generator[Dict[str, List[numpy.array]], None, None]:
```

Comment:
Can we streamline the API a little here?  `yield_one_pass` is essentially the same thing as `__call__`, just with a specific termination point and it seems strange to me that it has a completely different name and calling convention.  Since `__call__` can take parameters, how about just adding a `n_epochs=None` parameter to `__call__` that defaults to iterating forever?